[Automation] java run fails for org.apache.maven:maven-plugin-descriptor:2.1.0

## java run on ubuntu-latest / JDK 25 / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: 25
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="2.1.0" ./gradlew clean javaTest -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/26036219276/job/76536247352
Last 300 lines of the log:
```console
    "
    to contain:
      ["<lifecycles>", "<lifecycle>", "<id>default</id>", "<goal>compile</goal>"]
    but could not find:
      ["<lifecycles>"]
 
        at org_apache_maven.maven_plugin_descriptor.Maven_plugin_descriptorTest.lifecycleMappingsCanBeWrittenReadAndMutated(Maven_plugin_descriptorTest.java:322)

8 tests completed, 1 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
14 actionable tasks: 4 executed, 10 up-to-date

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for org.apache.maven:maven-plugin-descriptor:2.0.6 failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 2.1.0 at 2026-05-18T13:38:23Z with exit code 1 after 5s
❌ ERROR: 'native-image run' for 2.1.0 failed with exit code 1.
SOFT-FAIL[native-image run][2.1.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="2.1.0" ./gradlew clean test -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"]
========================================================================================
 org.apache.maven:maven-plugin-descriptor:2.0.6:2.1.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 2.1.0 at 2026-05-18T13:38:23Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="2.1.0" ./gradlew clean compileTestJava -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 786ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE

> Task :compileTestJava
Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/src/test/java/org_apache_maven/maven_plugin_descriptor/Maven_plugin_descriptorTest.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 1s
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 2.1.0 at 2026-05-18T13:38:26Z with exit code 0 after 3s
========================================================================================
 org.apache.maven:maven-plugin-descriptor:2.0.6:2.1.0 stage java run
========================================================================================
Starting stage 'java run' for 2.1.0 at 2026-05-18T13:38:26Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="2.1.0" ./gradlew clean javaTest -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 759ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE

> Task :compileTestJava
Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/src/test/java/org_apache_maven/maven_plugin_descriptor/Maven_plugin_descriptorTest.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Maven_plugin_descriptorTest > lifecycleMappingsCanBeWrittenReadAndMutated() FAILED
    java.lang.AssertionError: 
    Expecting actual:
      "<?xml version="1.0" encoding="UTF-8"?>
    <lifecycles xsi:schemaLocation="http://maven.apache.org/LIFECYCLE/1.0.0 http://maven.apache.org/xsd/lifecycle-1.0.0.xsd" xmlns="http://maven.apache.org/LIFECYCLE/1.0.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
      <lifecycle>
        <id>default</id>
        <phases>
          <phase>
            <id>process-sources</id>
            <executions>
              <execution>
                <configuration>
                  <skip>false</skip>
                </configuration>
                <goals>
                  <goal>compile</goal>
                </goals>
              </execution>
            </executions>
          </phase>
        </phases>
      </lifecycle>
    </lifecycles>
    "
    to contain:
      ["<lifecycles>", "<lifecycle>", "<id>default</id>", "<goal>compile</goal>"]
    but could not find:
      ["<lifecycles>"]
 
        at org_apache_maven.maven_plugin_descriptor.Maven_plugin_descriptorTest.lifecycleMappingsCanBeWrittenReadAndMutated(Maven_plugin_descriptorTest.java:322)

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

8 tests completed, 1 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s
13 actionable tasks: 3 executed, 10 up-to-date

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 2.1.0 at 2026-05-18T13:38:30Z with exit code 1 after 4s
❌ ERROR: 'java run' for 2.1.0 failed with exit code 1.
FAILED[java run][2.1.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="2.1.0" ./gradlew clean javaTest -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"]
```

## java run on ubuntu-latest / JDK latest-ea / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="2.1.0" ./gradlew clean javaTest -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/26036219276/job/76536247297
Last 300 lines of the log:
```console
    "
    to contain:
      ["<lifecycles>", "<lifecycle>", "<id>default</id>", "<goal>compile</goal>"]
    but could not find:
      ["<lifecycles>"]
 
        at org_apache_maven.maven_plugin_descriptor.Maven_plugin_descriptorTest.lifecycleMappingsCanBeWrittenReadAndMutated(Maven_plugin_descriptorTest.java:322)

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

8 tests completed, 1 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
14 actionable tasks: 4 executed, 10 up-to-date

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for org.apache.maven:maven-plugin-descriptor:2.0.6 failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 2.1.0 at 2026-05-18T13:37:45Z with exit code 1 after 5s
❌ ERROR: 'native-image run' for 2.1.0 failed with exit code 1.
SOFT-FAIL[native-image run][2.1.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="2.1.0" ./gradlew clean test -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"]
========================================================================================
 org.apache.maven:maven-plugin-descriptor:2.0.6:2.1.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 2.1.0 at 2026-05-18T13:37:45Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="2.1.0" ./gradlew clean compileTestJava -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 767ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE

> Task :compileTestJava
Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/src/test/java/org_apache_maven/maven_plugin_descriptor/Maven_plugin_descriptorTest.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 983ms
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 2.1.0 at 2026-05-18T13:37:48Z with exit code 0 after 3s
========================================================================================
 org.apache.maven:maven-plugin-descriptor:2.0.6:2.1.0 stage java run
========================================================================================
Starting stage 'java run' for 2.1.0 at 2026-05-18T13:37:48Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="2.1.0" ./gradlew clean javaTest -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 757ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :processTestResources NO-SOURCE
> Task :tck-test-support:jar UP-TO-DATE

> Task :compileTestJava

> Task :testClasses
Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/src/test/java/org_apache_maven/maven_plugin_descriptor/Maven_plugin_descriptorTest.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :test FAILED

Maven_plugin_descriptorTest > lifecycleMappingsCanBeWrittenReadAndMutated() FAILED
    java.lang.AssertionError: 
    Expecting actual:
      "<?xml version="1.0" encoding="UTF-8"?>
    <lifecycles xsi:schemaLocation="http://maven.apache.org/LIFECYCLE/1.0.0 http://maven.apache.org/xsd/lifecycle-1.0.0.xsd" xmlns="http://maven.apache.org/LIFECYCLE/1.0.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
      <lifecycle>
        <id>default</id>
        <phases>
          <phase>
            <id>process-sources</id>
            <executions>
              <execution>
                <configuration>
                  <skip>false</skip>
                </configuration>
                <goals>
                  <goal>compile</goal>
                </goals>
              </execution>
            </executions>
          </phase>
        </phases>

8 tests completed, 1 failed      </lifecycle>
    </lifecycles>
    "
    to contain:
      ["<lifecycles>", "<lifecycle>", "<id>default</id>", "<goal>compile</goal>"]
    but could not find:
      ["<lifecycles>"]
 
        at org_apache_maven.maven_plugin_descriptor.Maven_plugin_descriptorTest.lifecycleMappingsCanBeWrittenReadAndMutated(Maven_plugin_descriptorTest.java:322)



[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/tests/test/index.html

* Try:
> 
Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date
Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 3s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 2.1.0 at 2026-05-18T13:37:52Z with exit code 1 after 4s
❌ ERROR: 'java run' for 2.1.0 failed with exit code 1.
FAILED[java run][2.1.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="2.1.0" ./gradlew clean javaTest -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"]
```

## java run on ubuntu-latest / JDK latest-ea / mode future-defaults-all

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: future-defaults-all
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="2.1.0" ./gradlew clean javaTest -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/26036219276/job/76536247398
Last 300 lines of the log:
```console
    "
    to contain:
      ["<lifecycles>", "<lifecycle>", "<id>default</id>", "<goal>compile</goal>"]
    but could not find:
      ["<lifecycles>"]
 
        at org_apache_maven.maven_plugin_descriptor.Maven_plugin_descriptorTest.lifecycleMappingsCanBeWrittenReadAndMutated(Maven_plugin_descriptorTest.java:322)

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

8 tests completed, 1 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
14 actionable tasks: 4 executed, 10 up-to-date

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for org.apache.maven:maven-plugin-descriptor:2.0.6 failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 2.1.0 at 2026-05-18T13:38:16Z with exit code 1 after 4s
❌ ERROR: 'native-image run' for 2.1.0 failed with exit code 1.
SOFT-FAIL[native-image run][2.1.0][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="2.1.0" ./gradlew clean test -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"]
========================================================================================
 org.apache.maven:maven-plugin-descriptor:2.0.6:2.1.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 2.1.0 at 2026-05-18T13:38:16Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="2.1.0" ./gradlew clean compileTestJava -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 748ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE

> Task :compileTestJava

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 982ms
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html
Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/src/test/java/org_apache_maven/maven_plugin_descriptor/Maven_plugin_descriptorTest.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 2.1.0 at 2026-05-18T13:38:19Z with exit code 0 after 3s
========================================================================================
 org.apache.maven:maven-plugin-descriptor:2.0.6:2.1.0 stage java run
========================================================================================
Starting stage 'java run' for 2.1.0 at 2026-05-18T13:38:19Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="2.1.0" ./gradlew clean javaTest -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 758ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :checkTestTimeoutAnnotations
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE

> Task :compileTestJava

Note: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/src/test/java/org_apache_maven/maven_plugin_descriptor/Maven_plugin_descriptorTest.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Maven_plugin_descriptorTest > lifecycleMappingsCanBeWrittenReadAndMutated() FAILED
    java.lang.AssertionError: 
    Expecting actual:
      "<?xml version="1.0" encoding="UTF-8"?>
    <lifecycles xsi:schemaLocation="http://maven.apache.org/LIFECYCLE/1.0.0 http://maven.apache.org/xsd/lifecycle-1.0.0.xsd" xmlns="http://maven.apache.org/LIFECYCLE/1.0.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
      <lifecycle>
        <id>default</id>
        <phases>
          <phase>
            <id>process-sources</id>
            <executions>
              <execution>
                <configuration>
                  <skip>false</skip>
                </configuration>
                <goals>
                  <goal>compile</goal>
                </goals>
              </execution>
            </executions>
          </phase>
        </phases>
      </lifecycle>
    </lifecycles>
    "
    to contain:
      ["<lifecycles>", "<lifecycle>", "<id>default</id>", "<goal>compile</goal>"]
    but could not find:
      ["<lifecycles>"]
 
        at org_apache_maven.maven_plugin_descriptor.Maven_plugin_descriptorTest.lifecycleMappingsCanBeWrittenReadAndMutated(Maven_plugin_descriptorTest.java:322)

8 tests completed, 1 failed

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.apache.maven/maven-plugin-descriptor/2.0.6/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 2.1.0 at 2026-05-18T13:38:23Z with exit code 1 after 4s
❌ ERROR: 'java run' for 2.1.0 failed with exit code 1.
FAILED[java run][2.1.0][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="2.1.0" ./gradlew clean javaTest -Pcoordinates="org.apache.maven:maven-plugin-descriptor:2.0.6"]
```

--- a/metadata/org.apache.maven/maven-plugin-descriptor/2.1.0/reachability-metadata.json
+++ b/metadata/org.apache.maven/maven-plugin-descriptor/2.1.0/reachability-metadata.json
@@ -1,0 +1,10 @@
+{
+  "resources": [
+    {
+      "condition": {
+        "typeReached": "org.apache.maven.plugin.descriptor.PluginDescriptor"
+      },
+      "glob": "META-INF/maven/lifecycle.xml"
+    }
+  ]
+}

--- a/metadata/org.apache.maven/maven-plugin-descriptor/index.json
+++ b/metadata/org.apache.maven/maven-plugin-descriptor/index.json
@@ -2,6 +2,11 @@
   {
     "latest" : true,
     "metadata-version" : "2.1.0",
+    "source-code-url" : "https://repo.maven.apache.org/maven2/org/apache/maven/maven-plugin-descriptor/$version$/maven-plugin-descriptor-$version$-sources.jar",
+    "repository-url" : "https://svn.apache.org/repos/asf/maven/components",
+    "test-code-url" : "https://svn.apache.org/repos/asf/maven/components/tags/maven-$version$/maven-plugin-descriptor/src/test",
+    "documentation-url" : "https://repo.maven.apache.org/maven2/org/apache/maven/maven-plugin-descriptor/$version$/maven-plugin-descriptor-$version$-javadoc.jar",
+    "description" : "Apache Maven Plugin Descriptor provides the Maven 2 model classes and XML parser support for reading plugin descriptors that describe Mojos, parameters, requirements, dependencies, and plugin identity. It also includes the lifecycle mapping model and XPP3 readers and writers used to parse and serialize custom plugin lifecycle configuration.",
     "tested-versions" : [
       "2.1.0"
     ],

--- a/metadata/org.apache.maven/maven-plugin-descriptor/index.json
+++ b/metadata/org.apache.maven/maven-plugin-descriptor/index.json
@@ -1,6 +1,15 @@
 [
   {
     "latest" : true,
+    "metadata-version" : "2.1.0",
+    "tested-versions" : [
+      "2.1.0"
+    ],
+    "allowed-packages" : [
+      "org.apache.maven.plugin"
+    ]
+  },
+  {
     "metadata-version" : "2.0.6",
     "source-code-url" : "https://repo.maven.apache.org/maven2/org/apache/maven/maven-plugin-descriptor/$version$/maven-plugin-descriptor-$version$-sources.jar",
     "repository-url" : "https://svn.apache.org/repos/asf/maven/components",

--- a/stats/org.apache.maven/maven-plugin-descriptor/2.1.0/execution-metrics.json
+++ b/stats/org.apache.maven/maven-plugin-descriptor/2.1.0/execution-metrics.json
@@ -1,0 +1,93 @@
+{
+  "fix_java_run_fail:2026-05-18": {
+    "timestamp": "2026-05-18T22:19:55.258455Z",
+    "library": "org.apache.maven:maven-plugin-descriptor:2.1.0",
+    "starting_commit": "58c5c3605f9733d87ffe46be86e8b3ca9fad4d25",
+    "ending_commit": "d5f30159cd1ba11663632e1ff820d345a92b3477",
+    "previous_library": "org.apache.maven:maven-plugin-descriptor:2.0.6",
+    "strategy_name": "java_run_iterative_with_coverage_sources_pi_gpt-5.5",
+    "agent": "pi",
+    "model": "oca/gpt-5.5",
+    "status": "success",
+    "stats": {
+      "version": "2.1.0",
+      "dynamicAccess": {
+        "breakdown": {
+          "reflection": {
+            "coverageRatio": 1.0,
+            "coveredCalls": 4,
+            "totalCalls": 4
+          }
+        },
+        "coverageRatio": 1.0,
+        "coveredCalls": 4,
+        "totalCalls": 4
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 3182,
+          "missed": 696,
+          "ratio": 0.820526,
+          "total": 3878
+        },
+        "line": {
+          "covered": 799,
+          "missed": 134,
+          "ratio": 0.856377,
+          "total": 933
+        },
+        "method": {
+          "covered": 164,
+          "missed": 31,
+          "ratio": 0.841026,
+          "total": 195
+        }
+      }
+    },
+    "previous_library_stats": {
+      "version": "2.0.6",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 3122,
+          "missed": 534,
+          "ratio": 0.853939,
+          "total": 3656
+        },
+        "line": {
+          "covered": 814,
+          "missed": 112,
+          "ratio": 0.87905,
+          "total": 926
+        },
+        "method": {
+          "covered": 165,
+          "missed": 27,
+          "ratio": 0.859375,
+          "total": 192
+        }
+      }
+    },
+    "metrics": {
+      "input_tokens_used": 91068,
+      "cached_input_tokens_used": 381952,
+      "output_tokens_used": 5956,
+      "iterations": 5,
+      "cost_usd": 0.3697,
+      "tested_library_loc": 799,
+      "metadata_entries": 1,
+      "previous_library_metadata_entries": 0,
+      "code_coverage_percent": 85.64,
+      "previous_library_coverage_percent": 87.91
+    },
+    "artifacts": {
+      "test_file": "tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/LifecycleConfigurationTest.java",
+      "metadata_file": "metadata/org.apache.maven/maven-plugin-descriptor/2.1.0/reachability-metadata.json"
+    }
+  }
+}

--- a/stats/org.apache.maven/maven-plugin-descriptor/2.1.0/stats.json
+++ b/stats/org.apache.maven/maven-plugin-descriptor/2.1.0/stats.json
@@ -1,0 +1,37 @@
+{
+  "versions" : [ {
+    "version" : "2.1.0",
+    "dynamicAccess" : {
+      "breakdown" : {
+        "reflection" : {
+          "coverageRatio" : 1.0,
+          "coveredCalls" : 4,
+          "totalCalls" : 4
+        }
+      },
+      "coverageRatio" : 1.0,
+      "coveredCalls" : 4,
+      "totalCalls" : 4
+    },
+    "libraryCoverage" : {
+      "instruction" : {
+        "covered" : 3182,
+        "missed" : 696,
+        "ratio" : 0.820526,
+        "total" : 3878
+      },
+      "line" : {
+        "covered" : 799,
+        "missed" : 134,
+        "ratio" : 0.856377,
+        "total" : 933
+      },
+      "method" : {
+        "covered" : 164,
+        "missed" : 31,
+        "ratio" : 0.841026,
+        "total" : 195
+      }
+    }
+  } ]
+}

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/.gitignore
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/.gitignore
@@ -1,0 +1,4 @@
+gradlew.bat
+gradlew
+gradle/
+build/

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/build.gradle
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/build.gradle
@@ -1,0 +1,27 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+plugins {
+    id "org.graalvm.internal.tck"
+}
+
+String libraryVersion = tck.testedLibraryVersion.get()
+
+dependencies {
+    testImplementation "org.apache.maven:maven-plugin-descriptor:$libraryVersion"
+    testImplementation 'org.assertj:assertj-core:3.22.0'
+}
+
+graalvmNative {
+    agent {
+        defaultMode = "conditional"
+        modes {
+            conditional {
+                userCodeFilterPath = "user-code-filter.json"
+            }
+        }
+    }
+}

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/gradle.properties
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/gradle.properties
@@ -1,0 +1,6 @@
+# Optional explicit tested library coordinates (format: group:artifact:version).
+# If omitted, fallback resolution in org.graalvm.internal.tck.gradle still uses
+# environment variable GVM_TCK_LC and then this property.
+library.coordinates = org.apache.maven:maven-plugin-descriptor:2.1.0
+library.version = 2.1.0
+metadata.dir = org.apache.maven/maven-plugin-descriptor/2.1.0/

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/settings.gradle
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/settings.gradle
@@ -1,0 +1,20 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+pluginManagement {
+    def tckPath = Objects.requireNonNullElse(
+            System.getenv("GVM_TCK_TCKDIR"),
+            "../../../../tck-build-logic"
+    )
+    includeBuild(tckPath)
+    repositories { mavenLocal(); gradlePluginPortal() }
+}
+
+plugins {
+    id "org.graalvm.internal.tck-settings" version "1.0.0-SNAPSHOT"
+}
+
+rootProject.name = 'org.apache.maven.maven-plugin-descriptor_tests'

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/ExecutionTest.java
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/ExecutionTest.java
@@ -1,0 +1,25 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package org_apache_maven.maven_plugin_descriptor;
+
+import org.apache.maven.plugin.lifecycle.Execution;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class ExecutionTest {
+    @Test
+    void rejectsNullGoalWhenAddingGoal() {
+        Execution execution = new Execution();
+
+        ClassCastException exception = assertThrows(ClassCastException.class, () -> execution.addGoal(null));
+
+        assertThat(exception).hasMessageContaining("java.lang.String");
+        assertThat(execution.getGoals()).isEmpty();
+    }
+}

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/LifecycleConfigurationTest.java
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/LifecycleConfigurationTest.java
@@ -1,0 +1,26 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package org_apache_maven.maven_plugin_descriptor;
+
+import org.apache.maven.plugin.lifecycle.Lifecycle;
+import org.apache.maven.plugin.lifecycle.LifecycleConfiguration;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class LifecycleConfigurationTest {
+    @Test
+    void rejectsNullLifecycleWhenAddingLifecycle() {
+        LifecycleConfiguration configuration = new LifecycleConfiguration();
+
+        ClassCastException exception = assertThrows(ClassCastException.class, () -> configuration.addLifecycle(null));
+
+        assertThat(exception).hasMessageContaining(Lifecycle.class.getName());
+        assertThat(configuration.getLifecycles()).isEmpty();
+    }
+}

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/LifecycleTest.java
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/LifecycleTest.java
@@ -1,0 +1,26 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package org_apache_maven.maven_plugin_descriptor;
+
+import org.apache.maven.plugin.lifecycle.Lifecycle;
+import org.apache.maven.plugin.lifecycle.Phase;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class LifecycleTest {
+    @Test
+    void rejectsNullPhaseWhenAddingPhase() {
+        Lifecycle lifecycle = new Lifecycle();
+
+        ClassCastException exception = assertThrows(ClassCastException.class, () -> lifecycle.addPhase(null));
+
+        assertThat(exception).hasMessageContaining(Phase.class.getName());
+        assertThat(lifecycle.getPhases()).isEmpty();
+    }
+}

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/Maven_plugin_descriptorTest.java
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/Maven_plugin_descriptorTest.java
@@ -319,7 +319,7 @@ public class Maven_plugin_descriptorTest {
         new LifecycleMappingsXpp3Writer().write(writer, configuration);
 
         String xml = writer.toString();
-        assertThat(xml).contains("<lifecycles>", "<lifecycle>", "<id>default</id>", "<goal>compile</goal>");
+        assertThat(xml).contains("<lifecycles", "<lifecycle>", "<id>default</id>", "<goal>compile</goal>");
 
         LifecycleConfiguration parsed = new LifecycleMappingsXpp3Reader().read(new StringReader(xml));
         assertThat(parsed.getLifecycles()).hasSize(1);

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/Maven_plugin_descriptorTest.java
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/Maven_plugin_descriptorTest.java
@@ -1,0 +1,465 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package org_apache_maven.maven_plugin_descriptor;
+
+import java.io.FileNotFoundException;
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.artifact.DefaultArtifact;
+import org.apache.maven.artifact.handler.DefaultArtifactHandler;
+import org.apache.maven.artifact.versioning.VersionRange;
+import org.apache.maven.plugin.descriptor.DuplicateMojoDescriptorException;
+import org.apache.maven.plugin.descriptor.DuplicateParameterException;
+import org.apache.maven.plugin.descriptor.MojoDescriptor;
+import org.apache.maven.plugin.descriptor.Parameter;
+import org.apache.maven.plugin.descriptor.PluginDescriptor;
+import org.apache.maven.plugin.descriptor.PluginDescriptorBuilder;
+import org.apache.maven.plugin.lifecycle.Execution;
+import org.apache.maven.plugin.lifecycle.Lifecycle;
+import org.apache.maven.plugin.lifecycle.LifecycleConfiguration;
+import org.apache.maven.plugin.lifecycle.Phase;
+import org.apache.maven.plugin.lifecycle.io.xpp3.LifecycleMappingsXpp3Reader;
+import org.apache.maven.plugin.lifecycle.io.xpp3.LifecycleMappingsXpp3Writer;
+import org.codehaus.classworlds.ClassRealm;
+import org.codehaus.classworlds.ClassWorld;
+import org.codehaus.plexus.component.repository.ComponentDependency;
+import org.codehaus.plexus.component.repository.ComponentRequirement;
+import org.codehaus.plexus.configuration.PlexusConfiguration;
+import org.codehaus.plexus.configuration.PlexusConfigurationException;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class Maven_plugin_descriptorTest {
+    @TempDir
+    Path temporaryDirectory;
+
+    @Test
+    void buildsCompletePluginDescriptorFromXml() throws PlexusConfigurationException {
+        PluginDescriptor descriptor = new PluginDescriptorBuilder().build(new StringReader("""
+                <plugin>
+                  <groupId>com.example.build</groupId>
+                  <artifactId>sample-maven-plugin</artifactId>
+                  <version>1.0</version>
+                  <goalPrefix>sample</goalPrefix>
+                  <name>Sample Plugin</name>
+                  <description>Builds a sample project</description>
+                  <isolatedRealm>true</isolatedRealm>
+                  <inheritedByDefault>false</inheritedByDefault>
+                  <mojos>
+                    <mojo>
+                      <goal>compile</goal>
+                      <implementation>com.example.build.CompileMojo</implementation>
+                      <language>java</language>
+                      <configurator>basic</configurator>
+                      <composer>map-oriented</composer>
+                      <since>1.0</since>
+                      <phase>compile</phase>
+                      <executePhase>generate-sources</executePhase>
+                      <executeGoal>generate</executeGoal>
+                      <executeLifecycle>default</executeLifecycle>
+                      <instantiationStrategy>singleton</instantiationStrategy>
+                      <description>Compiles generated sources</description>
+                      <requiresDependencyResolution>test</requiresDependencyResolution>
+                      <requiresDirectInvocation>true</requiresDirectInvocation>
+                      <requiresProject>false</requiresProject>
+                      <requiresReports>true</requiresReports>
+                      <aggregator>true</aggregator>
+                      <requiresOnline>true</requiresOnline>
+                      <inheritedByDefault>false</inheritedByDefault>
+                      <parameters>
+                        <parameter>
+                          <name>sourceDirectory</name>
+                          <alias>sources</alias>
+                          <type>java.io.File</type>
+                          <required>true</required>
+                          <editable>false</editable>
+                          <description>Directory with sources</description>
+                          <deprecated>Use inputDirectory</deprecated>
+                          <implementation>java.io.File</implementation>
+                        </parameter>
+                      </parameters>
+                      <configuration>
+                        <sourceDirectory implementation="java.io.File">${source}</sourceDirectory>
+                      </configuration>
+                      <requirements>
+                        <requirement>
+                          <role>com.example.build.Compiler</role>
+                          <role-hint>javac</role-hint>
+                          <field-name>compiler</field-name>
+                        </requirement>
+                      </requirements>
+                    </mojo>
+                    <mojo>
+                      <goal>clean</goal>
+                      <implementation>com.example.build.CleanMojo</implementation>
+                      <description>Cleans generated output</description>
+                    </mojo>
+                  </mojos>
+                  <dependencies>
+                    <dependency>
+                      <groupId>com.example</groupId>
+                      <artifactId>build-support</artifactId>
+                      <type>jar</type>
+                      <version>2.1</version>
+                    </dependency>
+                  </dependencies>
+                </plugin>
+                """), "memory-plugin.xml");
+
+        assertThat(descriptor.getSource()).isEqualTo("memory-plugin.xml");
+        assertThat(descriptor.getGroupId()).isEqualTo("com.example.build");
+        assertThat(descriptor.getArtifactId()).isEqualTo("sample-maven-plugin");
+        assertThat(descriptor.getVersion()).isEqualTo("1.0");
+        assertThat(descriptor.getGoalPrefix()).isEqualTo("sample");
+        assertThat(descriptor.getName()).isEqualTo("Sample Plugin");
+        assertThat(descriptor.getDescription()).isEqualTo("Builds a sample project");
+        assertThat(descriptor.isIsolatedRealm()).isTrue();
+        assertThat(descriptor.isInheritedByDefault()).isFalse();
+        assertThat(descriptor.getId()).isEqualTo("com.example.build:sample-maven-plugin:1.0");
+        assertThat(descriptor.getPluginLookupKey()).isEqualTo("com.example.build:sample-maven-plugin");
+        assertThat(descriptor.getMojos()).hasSize(2);
+
+        MojoDescriptor mojo = descriptor.getMojo("compile");
+        assertThat(mojo.getPluginDescriptor()).isSameAs(descriptor);
+        assertThat(mojo.getGoal()).isEqualTo("compile");
+        assertThat(mojo.getFullGoalName()).isEqualTo("sample:compile");
+        assertThat(mojo.getImplementation()).isEqualTo("com.example.build.CompileMojo");
+        assertThat(mojo.getLanguage()).isEqualTo("java");
+        assertThat(mojo.getComponentConfigurator()).isEqualTo("basic");
+        assertThat(mojo.getComponentComposer()).isEqualTo("map-oriented");
+        assertThat(mojo.getSince()).isEqualTo("1.0");
+        assertThat(mojo.getPhase()).isEqualTo("compile");
+        assertThat(mojo.getExecutePhase()).isEqualTo("generate-sources");
+        assertThat(mojo.getExecuteGoal()).isEqualTo("generate");
+        assertThat(mojo.getExecuteLifecycle()).isEqualTo("default");
+        assertThat(mojo.getInstantiationStrategy()).isEqualTo("singleton");
+        assertThat(mojo.getDescription()).isEqualTo("Compiles generated sources");
+        assertThat(mojo.isDependencyResolutionRequired()).isEqualTo("test");
+        assertThat(mojo.isDirectInvocationOnly()).isTrue();
+        assertThat(mojo.isProjectRequired()).isFalse();
+        assertThat(mojo.isRequiresReports()).isTrue();
+        assertThat(mojo.isAggregator()).isTrue();
+        assertThat(mojo.requiresOnline()).isTrue();
+        assertThat(mojo.isInheritedByDefault()).isFalse();
+        assertThat(mojo.getParameterMap()).containsKey("sourceDirectory");
+
+        Parameter parameter = (Parameter) mojo.getParameters().get(0);
+        assertThat(parameter.getName()).isEqualTo("sourceDirectory");
+        assertThat(parameter.getAlias()).isEqualTo("sources");
+        assertThat(parameter.getType()).isEqualTo("java.io.File");
+        assertThat(parameter.isRequired()).isTrue();
+        assertThat(parameter.isEditable()).isFalse();
+        assertThat(parameter.getDescription()).isEqualTo("Directory with sources");
+        assertThat(parameter.getDeprecated()).isEqualTo("Use inputDirectory");
+        assertThat(parameter.getImplementation()).isEqualTo("java.io.File");
+        assertThat(parameter.toString()).contains("sourceDirectory", "sources");
+
+        PlexusConfiguration sourceDirectory = mojo.getMojoConfiguration().getChild("sourceDirectory");
+        assertThat(sourceDirectory.getValue()).isEqualTo("${source}");
+        assertThat(sourceDirectory.getAttribute("implementation")).isEqualTo("java.io.File");
+
+        ComponentRequirement requirement = (ComponentRequirement) mojo.getRequirements().get(0);
+        assertThat(requirement.getRole()).isEqualTo("com.example.build.Compiler");
+        assertThat(requirement.getRoleHint()).isEqualTo("javac");
+        assertThat(requirement.getFieldName()).isEqualTo("compiler");
+
+        ComponentDependency dependency = (ComponentDependency) descriptor.getDependencies().get(0);
+        assertThat(dependency.getGroupId()).isEqualTo("com.example");
+        assertThat(dependency.getArtifactId()).isEqualTo("build-support");
+        assertThat(dependency.getType()).isEqualTo("jar");
+        assertThat(dependency.getVersion()).isEqualTo("2.1");
+        assertThat(descriptor.getMojo("clean").getImplementation()).isEqualTo("com.example.build.CleanMojo");
+        assertThat(descriptor.getMojo("missing")).isNull();
+    }
+
+    @Test
+    void descriptorModelsExposeKeysDefaultsAndDuplicateValidation() throws Exception {
+        assertThat(PluginDescriptor.constructPluginKey("group", "artifact", "1"))
+                .isEqualTo("group:artifact:1");
+        assertThat(PluginDescriptor.getDefaultPluginGroupId()).isEqualTo("org.apache.maven.plugins");
+        assertThat(PluginDescriptor.getDefaultPluginArtifactId("help")).isEqualTo("maven-help-plugin");
+        assertThat(PluginDescriptor.getGoalPrefixFromArtifactId("maven-plugin-plugin")).isEqualTo("plugin");
+        assertThat(PluginDescriptor.getGoalPrefixFromArtifactId("maven-compiler-plugin")).isEqualTo("compiler");
+        assertThat(PluginDescriptor.getGoalPrefixFromArtifactId("sample-plugin-extension"))
+                .isEqualTo("sampleextension");
+
+        PluginDescriptor descriptor = new PluginDescriptor();
+        descriptor.setGroupId("com.example");
+        descriptor.setArtifactId("demo-plugin");
+        descriptor.setVersion("1.0");
+        descriptor.setGoalPrefix("demo");
+        descriptor.setSource("manual");
+        descriptor.setInheritedByDefault(true);
+        descriptor.setArtifacts(Collections.emptyList());
+
+        MojoDescriptor mojo = new MojoDescriptor();
+        mojo.setPluginDescriptor(descriptor);
+        mojo.setGoal("run");
+        mojo.setImplementation("com.example.RunMojo");
+        mojo.setDependencyResolutionRequired("runtime");
+        mojo.setExecutionStrategy("always");
+        mojo.setProjectRequired(true);
+        mojo.setOnlineRequired(false);
+        mojo.setAggregator(false);
+        mojo.setDirectInvocationOnly(false);
+        mojo.setRequiresReports(false);
+        descriptor.addMojo(mojo);
+
+        assertThat(mojo.getId()).isEqualTo("com.example:demo-plugin:1.0:run");
+        assertThat(mojo.getRole()).isEqualTo("org.apache.maven.plugin.Mojo");
+        assertThat(mojo.getRoleHint()).isEqualTo(mojo.getId());
+        assertThat(mojo.getComponentType()).isEqualTo("maven-plugin");
+        assertThat(mojo.alwaysExecute()).isTrue();
+        assertThat(mojo.isDependencyResolutionRequired()).isEqualTo("runtime");
+        assertThat(descriptor.getArtifacts()).isEmpty();
+        assertThat(descriptor.getArtifactMap()).isEmpty();
+
+        MojoDescriptor duplicateMojo = new MojoDescriptor();
+        duplicateMojo.setPluginDescriptor(descriptor);
+        duplicateMojo.setGoal("run");
+        duplicateMojo.setImplementation("com.example.OtherRunMojo");
+        assertThrows(DuplicateMojoDescriptorException.class, () -> descriptor.addMojo(duplicateMojo));
+
+        Parameter parameter = new Parameter();
+        parameter.setName("outputDirectory");
+        parameter.setAlias("output");
+        parameter.setType("java.io.File");
+        parameter.setExpression("${project.build.directory}");
+        parameter.setDefaultValue("target");
+        parameter.setSince("1.0");
+        parameter.setRequirement(new org.apache.maven.plugin.descriptor.Requirement("com.example.Output", "default"));
+        mojo.addParameter(parameter);
+
+        assertThat(parameter.getExpression()).isEqualTo("${project.build.directory}");
+        assertThat(parameter.getDefaultValue()).isEqualTo("target");
+        assertThat(parameter.getSince()).isEqualTo("1.0");
+        assertThat(parameter.getRequirement().getRole()).isEqualTo("com.example.Output");
+        assertThat(parameter.getRequirement().getRoleHint()).isEqualTo("default");
+        assertThat(mojo.getParameterMap()).containsEntry("outputDirectory", parameter);
+
+        Parameter duplicateParameter = new Parameter();
+        duplicateParameter.setName("outputDirectory");
+        assertThrows(DuplicateParameterException.class, () -> mojo.addParameter(duplicateParameter));
+    }
+
+    @Test
+    void pluginDescriptorIndexesArtifactsAndTracksIntroducedDependencies() {
+        Artifact alphaArtifact = new DefaultArtifact("com.example", "alpha-plugin", VersionRange.createFromVersion("1"),
+                Artifact.SCOPE_RUNTIME, "jar", null, new DefaultArtifactHandler("jar"));
+        Artifact betaArtifact = new DefaultArtifact("com.example", "beta-plugin", VersionRange.createFromVersion("1"),
+                Artifact.SCOPE_RUNTIME, "jar", null, new DefaultArtifactHandler("jar"));
+
+        PluginDescriptor descriptor = new PluginDescriptor();
+        descriptor.setArtifacts(Collections.singletonList(alphaArtifact));
+
+        Map artifactMap = descriptor.getArtifactMap();
+        assertThat(artifactMap).hasSize(1).containsEntry("com.example:alpha-plugin", alphaArtifact);
+        assertThat(descriptor.getIntroducedDependencyArtifacts()).isEmpty();
+
+        descriptor.setArtifacts(Collections.singletonList(betaArtifact));
+        assertThat(descriptor.getArtifactMap())
+                .hasSize(1)
+                .containsEntry("com.example:beta-plugin", betaArtifact)
+                .doesNotContainKey("com.example:alpha-plugin");
+
+        Set introducedDependencyArtifacts = new LinkedHashSet(Collections.singleton(betaArtifact));
+        descriptor.setIntroducedDependencyArtifacts(introducedDependencyArtifacts);
+        assertThat(descriptor.getIntroducedDependencyArtifacts()).containsExactly(betaArtifact);
+    }
+
+    @Test
+    void lifecycleMappingsCanBeWrittenReadAndMutated() throws Exception {
+        Xpp3Dom executionConfiguration = new Xpp3Dom("configuration");
+        Xpp3Dom skip = new Xpp3Dom("skip");
+        skip.setValue("false");
+        executionConfiguration.addChild(skip);
+
+        Execution execution = new Execution();
+        execution.addGoal("compile");
+        execution.addGoal("testCompile");
+        execution.removeGoal("testCompile");
+        execution.setConfiguration(executionConfiguration);
+
+        Phase phase = new Phase();
+        phase.setId("process-sources");
+        phase.addExecution(execution);
+
+        Lifecycle lifecycle = new Lifecycle();
+        lifecycle.setId("default");
+        lifecycle.addPhase(phase);
+
+        Lifecycle unusedLifecycle = new Lifecycle();
+        unusedLifecycle.setId("site");
+
+        LifecycleConfiguration configuration = new LifecycleConfiguration();
+        configuration.setModelEncoding("UTF-8");
+        configuration.addLifecycle(lifecycle);
+        configuration.addLifecycle(unusedLifecycle);
+        configuration.removeLifecycle(unusedLifecycle);
+
+        StringWriter writer = new StringWriter();
+        new LifecycleMappingsXpp3Writer().write(writer, configuration);
+
+        String xml = writer.toString();
+        assertThat(xml).contains("<lifecycles>", "<lifecycle>", "<id>default</id>", "<goal>compile</goal>");
+
+        LifecycleConfiguration parsed = new LifecycleMappingsXpp3Reader().read(new StringReader(xml));
+        assertThat(parsed.getLifecycles()).hasSize(1);
+
+        Lifecycle parsedLifecycle = (Lifecycle) parsed.getLifecycles().get(0);
+        assertThat(parsedLifecycle.getId()).isEqualTo("default");
+        assertThat(parsedLifecycle.getPhases()).hasSize(1);
+
+        Phase parsedPhase = (Phase) parsedLifecycle.getPhases().get(0);
+        assertThat(parsedPhase.getId()).isEqualTo("process-sources");
+        assertThat(parsedPhase.getExecutions()).hasSize(1);
+
+        Execution parsedExecution = (Execution) parsedPhase.getExecutions().get(0);
+        assertThat(parsedExecution.getGoals()).containsExactly("compile");
+        Xpp3Dom parsedExecutionConfiguration = (Xpp3Dom) parsedExecution.getConfiguration();
+        assertThat(parsedExecutionConfiguration.getChild("skip").getValue()).isEqualTo("false");
+
+        parsedPhase.removeExecution(parsedExecution);
+        assertThat(parsedPhase.getExecutions()).isEmpty();
+        parsedLifecycle.removePhase(parsedPhase);
+        assertThat(parsedLifecycle.getPhases()).isEmpty();
+    }
+
+    @Test
+    void readersAndBuildersRejectInvalidDescriptors() throws Exception {
+        PluginDescriptorBuilder builder = new PluginDescriptorBuilder();
+        assertThrows(PlexusConfigurationException.class,
+                () -> builder.buildConfiguration(new StringReader("<plugin>")));
+
+        String duplicateLifecycleId = """
+                <lifecycles>
+                  <lifecycle>
+                    <id>default</id>
+                    <id>duplicate</id>
+                  </lifecycle>
+                </lifecycles>
+                """;
+        assertThrows(XmlPullParserException.class,
+                () -> new LifecycleMappingsXpp3Reader().read(new StringReader(duplicateLifecycleId)));
+
+        String unknownLifecycleTag = """
+                <lifecycles>
+                  <lifecycle>
+                    <id>default</id>
+                    <unexpected />
+                  </lifecycle>
+                </lifecycles>
+                """;
+        assertThrows(XmlPullParserException.class,
+                () -> new LifecycleMappingsXpp3Reader().read(new StringReader(unknownLifecycleTag)));
+    }
+
+    @Test
+    void pluginDescriptorReportsMissingLifecycleResourceFromClassRealm() throws Exception {
+        PluginDescriptor descriptor = new PluginDescriptor();
+        ClassRealm realm = new ClassWorld().newRealm("empty-plugin-realm");
+        realm.addConstituent(temporaryDirectory.toUri().toURL());
+        descriptor.setClassRealm(realm);
+
+        assertThrows(FileNotFoundException.class, () -> descriptor.getLifecycleMapping("default"));
+    }
+
+    @Test
+    void pluginDescriptorLoadsLifecycleMappingsFromClassRealmResource() throws Exception {
+        Path lifecycleDirectory = Files.createDirectories(temporaryDirectory.resolve("META-INF/maven"));
+        Files.writeString(lifecycleDirectory.resolve("lifecycle.xml"), """
+                <lifecycles>
+                  <lifecycle>
+                    <id>default</id>
+                    <phases>
+                      <phase>
+                        <id>generate-resources</id>
+                        <executions>
+                          <execution>
+                            <goals>
+                              <goal>resources</goal>
+                            </goals>
+                            <configuration>
+                              <encoding>UTF-8</encoding>
+                            </configuration>
+                          </execution>
+                        </executions>
+                      </phase>
+                    </phases>
+                  </lifecycle>
+                  <lifecycle>
+                    <id>site</id>
+                    <phases>
+                      <phase>
+                        <id>site</id>
+                        <executions>
+                          <execution>
+                            <goals>
+                              <goal>site</goal>
+                            </goals>
+                          </execution>
+                        </executions>
+                      </phase>
+                    </phases>
+                  </lifecycle>
+                </lifecycles>
+                """);
+
+        PluginDescriptor descriptor = new PluginDescriptor();
+        ClassRealm realm = new ClassWorld().newRealm("plugin-lifecycle-realm");
+        realm.addConstituent(temporaryDirectory.toUri().toURL());
+        descriptor.setClassRealm(realm);
+
+        Lifecycle defaultLifecycle = descriptor.getLifecycleMapping("default");
+        assertThat(defaultLifecycle.getId()).isEqualTo("default");
+        Phase defaultPhase = (Phase) defaultLifecycle.getPhases().get(0);
+        assertThat(defaultPhase.getId()).isEqualTo("generate-resources");
+        Execution defaultExecution = (Execution) defaultPhase.getExecutions().get(0);
+        assertThat(defaultExecution.getGoals()).containsExactly("resources");
+        Xpp3Dom configuration = (Xpp3Dom) defaultExecution.getConfiguration();
+        assertThat(configuration.getChild("encoding").getValue()).isEqualTo("UTF-8");
+
+        Lifecycle siteLifecycle = descriptor.getLifecycleMapping("site");
+        assertThat(siteLifecycle.getId()).isEqualTo("site");
+        assertThat(descriptor.getLifecycleMapping("missing")).isNull();
+    }
+
+    @Test
+    void collectionSettersReplaceLifecycleModelContents() {
+        Execution execution = new Execution();
+        execution.setGoals(Arrays.asList("compile", "test"));
+        assertThat(execution.getGoals()).containsExactly("compile", "test");
+
+        Phase phase = new Phase();
+        phase.setExecutions(Collections.singletonList(execution));
+        phase.setConfiguration("phase-configuration");
+        assertThat(phase.getExecutions()).containsExactly(execution);
+        assertThat(phase.getConfiguration()).isEqualTo("phase-configuration");
+
+        Lifecycle lifecycle = new Lifecycle();
+        lifecycle.setPhases(Collections.singletonList(phase));
+        assertThat(lifecycle.getPhases()).containsExactly(phase);
+
+        LifecycleConfiguration configuration = new LifecycleConfiguration();
+        configuration.setLifecycles(Collections.singletonList(lifecycle));
+        assertThat(configuration.getLifecycles()).containsExactly(lifecycle);
+    }
+}

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/PhaseTest.java
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/src/test/java/org_apache_maven/maven_plugin_descriptor/PhaseTest.java
@@ -1,0 +1,26 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package org_apache_maven.maven_plugin_descriptor;
+
+import org.apache.maven.plugin.lifecycle.Execution;
+import org.apache.maven.plugin.lifecycle.Phase;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class PhaseTest {
+    @Test
+    void rejectsNullExecutionWhenAddingExecution() {
+        Phase phase = new Phase();
+
+        ClassCastException exception = assertThrows(ClassCastException.class, () -> phase.addExecution(null));
+
+        assertThat(exception).hasMessageContaining(Execution.class.getName());
+        assertThat(phase.getExecutions()).isEmpty();
+    }
+}

--- a/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/user-code-filter.json
+++ b/tests/src/org.apache.maven/maven-plugin-descriptor/2.1.0/user-code-filter.json
@@ -1,0 +1,13 @@
+{
+  "rules" : [
+    {
+      "excludeClasses" : "**"
+    },
+    {
+      "includeClasses" : "org.apache.maven.plugin.**"
+    },
+    {
+      "includeClasses" : "org_apache_maven.maven_plugin_descriptor.**"
+    }
+  ]
+}

